flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/parser/SqlClientSyntaxHighlighter.java

```diff
@@ -0,0 +1,256 @@
+     *
+     * <pre>
+     *      MultiLine Comment           Single Line Comment
+     *           |                              |
+     *   (&#47;*,*&#47;) |                              | (--, \n)
+     *           *------------Default-----------*
+     *                        |    |
+     *                        |    |
+     *           *------------*    *------------*
+     *  (&#47;*+,*&#47;) |            |    |            | (', ')
+     *           |            |    |            |
+     *         Hint           |    |          String
+     *                        |    |
+     *                        |    |
+     *           *------------*    *------------*
+     *    (&quot;, &quot;) |                              | (`, `)
+     *           |                              |
+     *     Hive Identifier           Flink Default Identifier
+     * </pre>
+     */
+    private enum State {
+        QUOTED(1, "'", "'", dialect -> true, (asb, style) -> asb.style(style.getQuotedStyle())),
+        SQL_QUOTED_IDENTIFIER(
+                2,
+                "`",
+                "`",
+                (dialect) -> dialect == SqlDialect.DEFAULT || dialect == null,
+                (asb, style) -> asb.style(style.getSqlIdentifierStyle())),
+        HIVE_SQL_QUOTED_IDENTIFIER(
+                2,
+                "\"",
+                "\"",
+                (dialect) -> dialect == SqlDialect.HIVE,
+                (asb, style) -> asb.style(style.getSqlIdentifierStyle())),
+        ONE_LINE_COMMENTED(
+                3, "--", "\n", dialect -> true, (asb, style) -> asb.style(style.getCommentStyle())),
+        HINTED(4, "/*+", "*/", dialect -> true, (asb, style) -> asb.style(style.getHintStyle())),
+        MULTILINE_COMMENTED(
+                5, "/*", "*/", dialect -> true, (asb, style) -> asb.style(style.getCommentStyle())),
+        DEFAULT(
+                Integer.MAX_VALUE,
+                null,
+                null,
+                dialect -> true,
+                (asb, style) -> asb.style(style.getDefaultStyle()));
+
+        private final String start;
+        private final String end;
+        private final Function<SqlDialect, Boolean> condition;
+
+        private final int order;
+
+        private final BiConsumer<AttributedStringBuilder, SyntaxHighlightStyle> styleSetter;
+
+        private static final List<State> STATE_LIST_WITHOUT_DEFAULT =
+                Arrays.stream(State.values())
+                        .filter(t -> t != DEFAULT)
+                        .sorted(Comparator.comparingInt(o -> o.order))
+                        .collect(Collectors.toList());
+        private static final Set<Character> STATE_START_SYMBOLS =
+                Arrays.stream(State.values())
+                        .filter(t -> t != DEFAULT)
+                        .map(t -> t.start.charAt(0))
+                        .collect(Collectors.toSet());
+
+        State(
+                int order,
```

Comment:
From one side yes, it makes sense...
From another side in that case I would add some javadoc mentioning that the order matters.
It would help if e.g. someone want to add more highlighting... e.g. highlight numbers separately